Thin host matrix wrapper

greole opened the pull request:
This PR adds a thin wrapper class around OpenFOAMs host LDU matrix class. The aim of this class is to simplify compute and access functions in order to create a ginkgo distributed matrix object. Ideally, this class doesn't need to be device persistent if we just make the distributed matrix class device persistent and an update function.  

Additionally it introduces a communication pattern wrapper class.

greole commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @param size how many elements to send to owner
+ * */
+CommCounts comput_scatter_from_owner_counts(const ExecutorHandler &exec_handler,
```
```suggestion
CommCounts compute_scatter_from_owner_counts(const ExecutorHandler &exec_handler,
```

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+/* @brief  This function computes the send and recv counts vectors for
+ * for scattering from an owner to non  owner ranks
+ *
+ * @param exec_handler The executor handler
```
What is executor handler?

greole replied:
The executor handler is responsible for persistent (over the course of a simulation) handling of ginkgo executor and communicator. It creates a ginkgo executor based on the "executor" entry in the dictionary. It computes corresponding device ids for allocation and checks if executor keywords are valid and the backend is compiled.

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+using CommCounts = std::tuple<std::vector<label>, std::vector<label>,
+                              std::vector<label>, std::vector<label>>;
+
+/* @brief  This function computes the send and recv counts vectors for
```
"for" is repeated in the next line

greole replied:
Please add a suggestion to remove the duplicate 'for'

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @param size how many elements to send to owner
+ * */
```
owner ranks are the MPI ranks which own the data after repartitioning */

greole replied:
correct.

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * for scattering from an owner to non  owner ranks
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
```
Ratio of the total number of ranks to the number of owner ranks

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @param size how many elements to send to owner
```
how many elements to send from owner

greole replied:
no to the owner. the comm counts are used for gathering to owner ranks.

chihtaw replied:
This block of comments should describe the function right below it. The function is `compute_scatter_from_owner_counts`, so I think it should compute the number of elements sent from owner.

greole replied:
Ah yes, you're right.

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @parameter size how many elements to send to owner
```
send from owner

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+CommCounts compute_scatter_from_owner_comm(const ExecutorHandler &exec_handler,
+                                           label ranks_per_owner, label size);
+
+/** @brief Given a rank id and ranks_per_owners this function computes the
```
ranks_per_owner,

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ *
+ * @param rank The rank under consideration
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @return The corresponding owner of this ranks
```
The corresponding owner rank for a given rank

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * calls so that we can inspect the communication for debugging purposes
+ *
+ * @param exec_handler The executor handler
+ * @param comm_pattern The communication pattern / send recv counts and offsets
```
the communication pattern simply means "send recv count and offsets"?

greole replied:
yes.

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * */
+CommCounts compute_scatter_from_owner_comm(const ExecutorHandler &exec_handler,
+                                           label ranks_per_owner, label size);
+
```
This purpose of this function is similar to the purpose of another function `comput_scatter_from_owner_counts`
Why do we need these two similar functions?

greole replied:
Good catch, this function isn't needed.

chihtaw commented on src/CommunicationPattern/CommunicationPattern.H:
```diff
@@ -0,0 +1,189 @@
+ * @param exec_handler The executor handler
+ * @param send_buffer Pointer to send data
+ * @param send_size How many elements to send to owner
+ * @param offset Value to add to all elements in the send_buffer
```
offset here is not the displacement? but is the value to increment all elements in the send_buffer?

greole replied:
yes. This is needed for repartitioning the row and column indices and adapt the local indices by the rank offset.

chihtaw commented on unitTests/CommunicationPattern.C:
```diff
@@ -0,0 +1,108 @@
+    auto comm = exec->get_gko_mpi_device_comm();
+    auto owner_rank = compute_owner_rank(comm->rank(), comm->size());
+    // if ranks_per_owner is same as total number of ranks all ranks have
+    // zero as owner
```
```suggestion
    // when all ranks have the rank 0 as the owner rank
```

chihtaw commented on unitTests/CommunicationPattern.C:
```diff
@@ -0,0 +1,108 @@
+    EXPECT_EQ(std::get<0>(comm_counts), send_results[comm->rank()]);
+    EXPECT_EQ(std::get<1>(comm_counts), send_results[comm->rank()]);
+
+    // test send counts and revc offsets
```
Should be send offsets?

chihtaw commented on unitTests/CommunicationPattern.C:
```diff
@@ -0,0 +1,108 @@
+    EXPECT_EQ(std::get<0>(comm_counts), send_results[comm->rank()]);
+    EXPECT_EQ(std::get<1>(comm_counts), recv_results[comm->rank()]);
+
+    // test send counts and revc offsets
```
should be send offsets?

greole committed:
Wip communication pattern and hostmatrix wrapper impl

greole committed:
reorder library

greole committed:
reorder library

greole committed:
move implementation to .C file

greole committed:
rename file

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of the total number of ranks to owner ranks
+ * @param size how many elements to send to owner
```
According to the function name `compute_scatter_from_owner_counts`, the argument `size` should be the number of elements to be sent **from** owner

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+ * corresponding owner rank
+ *
+ * @param rank The rank under consideration
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
```
```suggestion
 * @param ranks_per_owner Ratio of the total number of ranks to owner ranks
```

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+    std::vector<label> recv_offsets;
+};
+
+/* @brief  This function computes the send and recv counts vectors for
```
```suggestion
/* @brief  This function computes the send and recv counts vectors and the send and recv offsets vectors for
```

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+    // to the corresponding target_id
+    gko::array<comm_size_type> target_sizes;
+
+    // send_idx stores the index_set ie which cells
```
Where is the variable `index_set` from?

chihtaw replied:
```suggestion
    // send_idxs stores the index_set ie which cells
```

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+                                          label send_size, label offset = 0);
+
+struct CommunicationPattern {
+    using comm_size_type = label;
```
`comm_size_type` is defined as a type alias of the type `label` here, but sometimes the type alias `comm_size_type` and the type `label` are used together in the following declarations, e.g., the declaration of the variable `send_idxs `. 
It causes some confusion.

greole replied:
Agree, i guess we can use label `everywhere`

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+
+    // send_idx stores the index_set ie which cells
+    // are owned by the interface and the corresponding target rank
+    std::vector<std::pair<gko::array<label>, comm_size_type>> send_idxs;
```
The type `label` and its type alias `comm_size_type` are used together. To avoid some confustion, it would be better if only one of them is used.

chihtaw commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+    CommunicationPattern(
+        const ExecutorHandler &exec, gko::array<comm_size_type> ids,
+        gko::array<comm_size_type> sizes,
+        std::vector<std::pair<gko::array<label>, comm_size_type>> idxs)
```
The type `label` and its type alias `comm_size_type` are used together. To avoid some confustion, it would be better if only one of them is used.

greole committed:
prepare host matrix unittest

greole committed:
add basic hostmatrix unit test

greole committed:
add basic hostmatrix unit test

greole committed:
Apply suggestions from code review

Co-authored-by: chihta <97195768+chihta-wang@users.noreply.github.com>

greole committed:
Merge branch 'impl/host_wrapper' of github.com:hpsim/OGL into impl/host_wrapper

greole committed:
deactivate failing test

greole committed:
fix failing test due to call to pure virtual function

greole committed:
format files

greole committed:
wip read test case from disk

greole commented on include/OGL/CommunicationPattern.H:
```diff
@@ -0,0 +1,148 @@
+                                          label send_size, label offset = 0);
+
+struct CommunicationPattern {
+    using comm_size_type = label;
```
```suggestion
```

greole replied:
```suggestion
```

chihtaw committed:
Rename the unit test for the function compute_owner_rank
to make it clear that there is only one owner process in the test

chihtaw committed:
Remove the unit test for the size of a MPI communicator

The unit test does not verify any functionality of CommunicationPattern.
It verifies the functionality of a MPI communicator, and thus should not
belong to the test suite for CommunicationPattern.

To make some hardcoded unit tests more self-contained, the ascertain part
for the communicator size is added accordinlgy.

chihtaw committed:
Add a unit test for the function compute_owner_rank for the case
of two owner processes

chihtaw committed:
Add a unit test for the function compute_owner_rank for the case:
All processes in the communicator are owner processes.

greole committed:
wip setup gtest

chihtaw committed:
Add the structure of unit testing to the test suite for CommunicationPattern
to make the test code more readable

chihtaw committed:
Parameterize the number of elements sent to owner processes by each process
to improve readability

chihtaw committed:
Adapt the unit test for the function compute_gather_to_owner_counts
for the case of only a single owner process so that any number of elements
can be set for the test

greole committed:
wip setup gtest, missing: copy of test case

chihtaw committed:
Refactor the unit test "compute_gather_to_owner_counts_all_owner"
In the arrange section,
1. remove the vector of vector previously used for expected results
   Use a vector for the expected result instead.
2. Add vectors for the expected results of receive parts (recv_counts and recv_offsets)

In the assert section,
use vectors instead of vectors of vectors for both send and receive parts

chihtaw committed:
Refactor the unit test "compute_gather_to_owner_counts_single_owner"

1. Use vectors instead of vectors of vectors for expected results.
2. Parameterize the communicator size to avoid the repeated function calls.

chihtaw committed:
Modify the unit test "compute_gather_to_owner_counts_single_owner" so that it is not
hardcoded anymore.
Now it can run on an arbitrarily even number of processes.

chihtaw committed:
Reorder the unit tests for the function "compute_gather_to_owner_counts"
in the increasing number of owner processes.

chihtaw committed:
Add a new unit test for the function "compute_gather_to_owner_counts"
for the case of two owner processes

greole committed:
fix link issue, add more unit tests

chihtaw committed:
Modify the unit tests "compute_gather_to_owner_counts_all_owners"
and "compute_gather_to_owner_counts_single_owner" to verify
that each process can send a different number of elements to owner processes.

greole committed:
add non local sparsity unit tests, clean-up, format

chihtaw committed:
Add a unit test for the function "compute_scatter_from_owner_counts"

chihtaw committed:
owner rank 0 does not send data to itself

chihtaw committed:
Comment out the unit test "compute_scatter_from_owner_counts_two_owners"
for the moment.

Modify the unit test "compute_scatter_from_owner_counts_single_owner" so that
the owner can send data to itself.

greole committed:
fix send/revc offsets back for scatter pattern

chihtaw committed:
Add a check for the number of CPU processes used to run the unit tests
If only 1 process is used, the program will abort.

chihtaw committed:
Add curly braces for if statement and for loop statement to make these structures
more easily identified.

chihtaw committed:
Add a new unit test for the function "compute_scatter_from_owner_counts" for
the case of two owners.

chihtaw committed:
Use clang-format to set the format style for the unit test code

chihtaw committed:
Correct the Doxygen comment for the function "compute_scatter_from_owner_counts"

greole committed:
fix compilation on mac

greole committed:
use test environment instead of fixture

greole committed:
Merge pull request #143 from hpsim/impl/unit_tests_communication-pattern

implement additional unit tests for communication pattern, see https://github.com/hpsim/OGL/pull/143

greole committed:
format files

greole committed:
format files

greole committed:
format files

greole committed:
fix typos

greole committed:
format files

greole committed:
format files

greole committed:
clean up

greole committed:
add test case

greole commented on unitTests/HostMatrix.C:
```diff
@@ -0,0 +1,215 @@
+    EXPECT_EQ(localSparsity->dim[1], 9);
+
+    // since we don't have any processor interfaces we only have
+    // a single interface span ranging from 0 to 17
```
```suggestion
    // a single interface span ranging from 0 to 33
```

greole committed:
Update unitTests/HostMatrix.C

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+ ** @param cols pointer to columns array
+ ** @param permute pointer to permuter array
+ */
+void init_local_sparsity(const label nrows, const label upper_nnz,
```
Why only upper non-zeros are considered? How about the lower non-zeros?

greole replied:
`const label upper_nnz` is just the number of entries in the upper triangular matrix. the number of upper_nnz==lower_nnz

greole replied:
```suggestion
void init_local_sparsity(const label nrows, const label non_diag_nnz,
```

greole replied:
i'll keep it as `upper_nnz`, because we already use `non_diag_nnz` for 2*upper_nnz

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+
+    const ExecutorHandler &exec_;
+
+    const DeviceIdGuardHandler device_id_guard_;
```
What is `device_id_guard`?

greole replied:
for more details see `scoped_device_id_guard.hpp` in Ginkgo

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+    ** ret = [(1,2,1),(2,20, 2), (3, 20, 2) ...]
+    **         i0   i1,   i...
+    */
+    [[nodiscard]] std::vector<std::tuple<label, label, label, label>>
```
There is a inconsistency between the comment and the function declaration regarding the return value.
In the comment, a vector of 3-element tuples is specified for the return value. However, the function declaration has a vector of 4-element tuples as the return type.

greole replied:
good catch.

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+        const lduInterfaceFieldPtrsList &interfaces) const;
+
+public:
+    // segregated matrix wrapper constructor
```
What is a segregated matrix?

greole replied:
OF uses the term segregated opposed to coupled matrix. In a coupled matrix all transport equations are stored in a single matrix. In a segregated approach each transport equation gets its own matrix.

chihtaw commented on unitTests/HostMatrix.C:
```diff
@@ -0,0 +1,212 @@
+    EXPECT_EQ(((HostMatrixEnvironment *)global_env)->args_->size(), 1);
+}
+
+TEST(HostMatrix, runsInParallelWithCorrectNumberOfRanks)
```
Since the unit tests for host matrix wrapper are hardcoded at the moment, it's better to put this test to a place where setup takes place, e.g., the body of the `SetUp` function

greole replied:
Aggree, will do that.

chihta-wang commented on unitTests/HostMatrix.C:
```diff
@@ -0,0 +1,212 @@
+        nonLocalSparsity->row_idxs.get_data() + res_size);
+    EXPECT_EQ(rows_expected[comm->rank()], rows_res);
+}
+
```
We may need some unit tests for the member functions of the class `HostMatrixWrapper`?

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+     ** This uses the local_sparsity_.ldu_mapping to permute the data already
+     ** on the host or device to be in row major order.
+     **/
+    void compute_local_coeffs(std::shared_ptr<const SparsityPattern> sparsity,
```
Is `compute_local_coeffs` implemented?

greole replied:
Currently, not. Will either added here in a separate PR or somewhere else. The reason is that it depends on the new distributed matrix wrapper class which is currently missing.

chihtaw commented on include/OGL/MatrixWrapper/HostMatrix.H:
```diff
@@ -0,0 +1,231 @@
+    /** Copies the interface matrix coefficients to non_local_coeffs without
+    ** changing or reinstantiating the sparsity pattern.
+    **/
+    void compute_non_local_coeffs(
```
Is `compute_non_local_coeffs` implemented?

greole replied:
Same as for the local_coeffs.

chihtaw reviewed (commented):
Please check the comments for the header file `HostMatrix.H` and for the unit test code  `HostMatrix.C`.

We may need some unit tests for the member functions of the class `HostMatrixWrapper`?

greole committed:
move non tests to env setup

greole committed:
format
